[BE] feat: 결제 인가 추가

Shin-Dong-Jun opened the pull request:
1. 결제 부분 인가 추가
2. 불필요 코드 삭제
3. Payment 결제 취소시 completedAt = cancledAt 부분 삭제(결제이력에서 관리하므로 불필요)

<!-- 제목: [BE/FE/All] (기능 등) -->
<!-- 아래에 이슈 번호를 매겨주세요 -->

- resolves #227 

---

### 🚀 어떤 기능을 구현했나요 ?

- 1. 결제 부분 인가 추가
2. 불필요 코드 삭제
3. Payment 결제 취소시 completedAt = cancledAt 부분 삭제(결제이력에서 관리하므로 불필요)

### 🔥 어떻게 해결했나요 ?
- 1. 결제 부분 인가 추가
2. 불필요 코드 삭제
3. Payment 결제 취소시 completedAt = cancledAt 부분 삭제(결제이력에서 관리하므로 불필요)

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 

### 📚 참고 자료, 할 말

Shin-Dong-Jun committed:
feat: 결제 인가 추가
1. 결제 부분 인가 추가
2. 불필요 코드 삭제
3. Payment 결제 취소시 completedAt = cancledAt 부분 삭제(결제이력에서 관리하므로 불필요)

Shin-Dong-Jun committed:
Merge branch 'dev' of https://github.com/tutti-tutti/tutti-server into be/feat/227-authorization-aop

heymeworld commented on src/main/java/com/tutti/server/core/payment/application/PaymentService.java:
Long memberId
Long AuthMemberId
이런 변수명은 맞춰주면 좋을 것 같아요

heymeworld commented on src/main/java/com/tutti/server/core/payment/application/PaymentViewServiceImpl.java:
```diff
@@ -20,7 +20,11 @@ public class PaymentViewServiceImpl implements PaymentViewService {
     @Override
     @Transactional(readOnly = true)
-    public List<PaymentViewResponse> viewPaymentByMemberId(Long memberId) {
+    public List<PaymentViewResponse> viewPaymentByMemberId(Long memberId, Long authMemberId) {
```
memberId 가 2개!

Shin-Dong-Jun replied:
아 요거는 첫번째 memberId는 Path 파라미터로 받은 memberId 이고 두번째는 컨트롤러에서  @AuthenticationPrincipal CustomUserDetails userDetails 여기서 꺼낸 memberId입니당.
혜미님 코드를 보면서도 Repo에서 찾는 형식으로 통일시키는것도 좋을것 같다는 생각을 하긴 했씁니당 밥먹고 와서 수정할겠습니당~~~

heymeworld reviewed (changes requested):
1차 리뷰 완료

전체적으로, payment.validateOwner(AuthMemberId); 메서드를 쓰는 것보다
애초에 Repository에서 payment를 찾을 때 이런 식으로 써주면 어때요?
`paymentReposutory.findByMemberIdAndIdAndDeleteStatusFalse(memberId, paymentId)`

orderRepository 쪽에도 memberId와 매칭시키는 메서드 이미 있습니다!

Shin-Dong-Jun reviewed (commented):
1차 답변완료! 

밥먹고 수정예정~~

Shin-Dong-Jun committed:
feat: payment 인가 메서드 -> orderRepository 메서드로 대체

Shin-Dong-Jun committed:
feat: payment 인가 메서드 -> orderRepository 메서드로 대체
1. 결제취소시간을 받을 필요가 없어짐에 따른 수정
2. 토스페이먼츠에서 에러가 발생했을 때 예외처리

Shin-Dong-Jun committed:
Merge branch 'dev' of https://github.com/tutti-tutti/tutti-server into be/feat/227-authorization-aop

# Conflicts:
#	src/main/java/com/tutti/server/core/support/exception/ExceptionCode.java

Shin-Dong-Jun committed:
feat: payment 인가 메서드 -> orderRepository 메서드로 대체
1. 결제취소시간을 받을 필요가 없어짐에 따른 수정
2. 토스페이먼츠에서 에러가 발생했을 때 예외처리
3. 인가처리완료
4. orderRepo에 쿼리메서드 추가(orderNumber와 authMemberId로 인가 체크를 위함

Shin-Dong-Jun reviewed (commented):
파일체인지가 좀 많긴 하네요... 죄송함다.. 삽질을 좀 했어서.. 

삽질 
1. order를 쿼리로 넣어놓고 테스트를 했는데 deleteStatus를 1로 해놨어서 계속 권한이 없다고 떴었음.. 뒤늦게 발견
2. memberId, orderId 둘다 Long이라 순서 바꿔썼었음.. 헤미님 노션덕분에 일찍 해결
3. payment/request 할때는 orderId(orderNumber)라서 String으로 된 쿼리메서드 orderRepo에 추가하였습니다~

heymeworld commented on src/main/java/com/tutti/server/core/payment/application/TossPaymentService.java:
```diff
@@ -79,14 +81,22 @@ public TossPaymentsCancelResponse cancelPayment(Payment payment, String cancelRe
+            if (response == null) {
+                throw new DomainException(ExceptionType.TOSS_NOT_RESPONSE);
+            }
+            return TossPaymentsCancelResponse.fromResponse(response);
```
이거는 TossPaymentsCancelResponse 를 TossPaymentsCancelResponse 엔티티군요.
from 메서드 없이 @Builder 붙여서 할 수 있을 것 같은데

'''
TossPaymentsCancelResponse.builder()
                .status(status)
                .canceledAt((List<T> ~~~~~~~
        )
'''

이렇게요!

heymeworld commented on src/main/java/com/tutti/server/core/payment/api/PaymentApi.java:
```diff
@@ -35,35 +39,42 @@ public class PaymentApi implements PaymentApiSpec {
     // 회원ID로 결제 조회
     @GetMapping("/memberId/{memberId}")
-    public List<PaymentViewResponse> getMemberPayments(@PathVariable Long memberId) {
+    public List<PaymentViewResponse> getMemberPayments(@PathVariable Long memberId,
+            @AuthenticationPrincipal CustomUserDetails user) {
```
결제를 조회할 때는 payment Id 를 쓰는 게 RESTful 한 것 같아요~
그리고 클라이언트한테 member Id 를 입력받으면 안 됩니다!

'''
        if (!memberId.equals(authMemberId)) {
            throw new DomainException(ExceptionType.UNAUTHORIZED_ERROR);
        }
'''

이 로직이 없어도 paymentRepository.findByIdAndMemberId(paymentId, memberId); 로 충분히 검증할 수 있을 것 같습니다~~

heymeworld reviewed (changes requested):
2차 리뷰 완.

payload에 request / response 로 패키지를 나눠주세요 !

Shin-Dong-Jun committed:
feat: 인가 수정
1. OrderRepository의 orderNumber로 확인할 수 있는 메서드 추가
2. TossPaymentService에 builder가 아닌 record로 바로 매핑이 가능하도록 변경
3. 기존 메서드말고 쿼리로 "권한 없음" 확인 할 수 있도록 변경
4. response, reqeust 패키지 분리
5. memberId로 조회 기능 -> paymentId로 조회 기능으로 바꿈(List 조회(다건 조회) -> 단건 조회)

Shin-Dong-Jun committed:
Merge branch 'dev' of https://github.com/tutti-tutti/tutti-server into be/feat/227-authorization-aop

# Conflicts:
#	src/main/java/com/tutti/server/core/order/infrastructure/OrderRepository.java

Shin-Dong-Jun committed:
feat: orderRepository 매개변수 순서 변동에 대한 수정

Shin-Dong-Jun reviewed (commented):
헉.. response, request 패키지 나눴더니 확늘었네요... 10개가(DTO관련된건데 양해 부탁드립니당..


feat: 인가 수정
1. OrderRepository의 orderNumber로 확인할 수 있는 메서드 추가
2. TossPaymentService에 builder가 아닌 record로 바로 매핑이 가능하도록 변경
3. 기존 메서드말고 쿼리로 "권한 없음" 확인 할 수 있도록 변경
4. response, reqeust 패키지 분리
5. memberId로 조회 기능 -> paymentId로 조회 기능으로 바꿈(List 조회(다건 조회) -> 단건 조회)

feat: orderRepository 매개변수 순서 변동에 대한 수정

 변경사항은 위와 같습니다!

heymeworld commented on src/main/java/com/tutti/server/core/payment/payload/response/TossPaymentsCancelResponse.java:
```diff
@@ -0,0 +1,23 @@
+
+public record TossPaymentsCancelResponse(
+        String status,
+        List<CancelDetail> cancels
```
요 DTO의 용도가 궁금합니당
왜 List<CancelDetail> cancels 리스트를 받는지?

Shin-Dong-Jun replied:
토스에서 응답이 올 때 canceldAt 값이 중첩 json안에 있어서 저렇게 작성했어여
from메서드랑 builder를 안쓰고 DTO랑 바로 매핑시키려면 구조를 맞춰야한다고해서 저렇게 작성했습니다!

heymeworld replied:
그렇군요
그러면 값이 List 여서가 아니고, depth 때문이라면 이렇게 해도 되지 않나용??
```suggestion
        CancelDetail cancel
```

Shin-Dong-Jun replied:
토스에서 응답을 배열 형태로 줘서 저희는 부분취소가 없긴해도 토스에서 주는 응답형태를 그대로 따라야할 것 같아요

Shin-Dong-Jun committed:
Merge branch 'dev' of https://github.com/tutti-tutti/tutti-server into be/feat/227-authorization-aop

heymeworld commented on src/main/java/com/tutti/server/core/payment/api/PaymentApiSpec.java:
```diff
@@ -1,30 +1,30 @@
 
-    @Operation(summary = "회원 ID로 결제 조회", description = "특정 회원 ID를 기반으로 결제 내역을 조회하는 API")
-    List<PaymentViewResponse> getMemberPayments(Long memberId);
+    @Operation(summary = "paymentId로 결제 조회", description = "특정 회원 ID를 기반으로 결제 내역을 조회하는 API")
+    PaymentViewResponse getPaymentIdViewPayments(Long memberId, CustomUserDetails user);
```
여기 파라미터 안 바꼈네요~~

Shin-Dong-Jun replied:
수정했습니당

heymeworld commented:
3차 리뷰 완

Shin-Dong-Jun committed:
feat: PaymentApiSpec 파라미터 수정

Shin-Dong-Jun reviewed (commented):
답변 및 파라미터 수정하였습니다!

heymeworld reviewed (changes requested):
4차

Shin-Dong-Jun reviewed (commented):
답변 완료입니당
